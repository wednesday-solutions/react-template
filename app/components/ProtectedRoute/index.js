/**
 *
 * ProtectedRoute
 *
 */

import React from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import PropTypes from 'prop-types';
import routeConstants from '@utils/routeConstants';

const ProtectedRoute = ({ component: Component, isLoggedIn, handleLogout, ...rest }) => {
  const location = useLocation();

  const getUnprotectedRoutes = () =>
    Object.keys(routeConstants)
      .filter((key) => !routeConstants[key].isProtected)
      .map((key) => routeConstants[key].route);

  const isUnprotectedRoute = getUnprotectedRoutes().includes(location.pathname);

  const handleNotLoggedIn = () => {
    if (isUnprotectedRoute) {
      return <Component {...rest} />;
    }
    if (handleLogout) {
      handleLogout();
    }
    return <Navigate to={routeConstants.login?.route || '/'} replace />;
  };

  const handleLoggedIn = () => {
    if (isUnprotectedRoute) {
      return <Navigate to={routeConstants.dashboard?.route || '/'} replace />;
    }
    return <Component {...rest} />;
  };

  return isLoggedIn ? handleLoggedIn() : handleNotLoggedIn();
};

ProtectedRoute.propTypes = {
  component: PropTypes.any,
  isLoggedIn: PropTypes.bool,
  isUserVerified: PropTypes.bool,
  handleLogout: PropTypes.func
};

export default ProtectedRoute;
