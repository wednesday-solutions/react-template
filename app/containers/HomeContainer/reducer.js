/*
 *
 * HomeContainer reducer
 *
 */
import { produce } from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetGithubRepos: ['repoName'],
  successGetGithubRepos: ['data'],
  failureGetGithubRepos: ['error'],
  clearGithubRepos: {}
});
export const initialState = { repoName: null, reposData: {}, reposError: null, loading: null };

export const homeContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_GITHUB_REPOS:
        draft.repoName = action.repoName;
        draft.loading = true;
        break;
      case homeContainerTypes.CLEAR_GITHUB_REPOS:
        draft.repoName = null;
        draft.reposError = null;
        draft.reposData = {};
        draft.loading = null;
        break;
      case homeContainerTypes.SUCCESS_GET_GITHUB_REPOS:
        draft.reposData = action.data;
        draft.reposError = null;
        draft.loading = false;
        break;
      case homeContainerTypes.FAILURE_GET_GITHUB_REPOS:
        draft.reposError = get(action.error, 'message', 'something_went_wrong');
        draft.reposData = null;
        draft.loading = false;
        break;
    }
  });

export default homeContainerReducer;
