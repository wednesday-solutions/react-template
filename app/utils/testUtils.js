import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { browserHistory, BrowserRouter, Router } from 'react-router-dom';
import { ThemeProvider } from '@mui/material/styles';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import { IntlGlobalProvider } from '@components/IntlGlobalProvider';
import { theme } from '@containers/App';

export const renderWithIntl = (children) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <IntlGlobalProvider>{children}</IntlGlobalProvider>
    </IntlProvider>
  );

export const renderProvider = (children, history) => {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider theme={theme}>
          {history ? <Router history={history}>{children}</Router> : <BrowserRouter>{children}</BrowserRouter>}
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});

export const getStyles = () => {};
