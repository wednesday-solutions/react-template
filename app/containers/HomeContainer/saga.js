import { put, call, takeLatest } from 'redux-saga/effects';
import { getRepos } from '@services/repoApi';
import { homeContainerTypes, homeContainerCreators } from './reducer';

const { REQUEST_GET_GITHUB_REPOS } = homeContainerTypes;
const { successGetGithubRepos, failureGetGithubRepos } = homeContainerCreators;

/**
 * A saga that handles fetching GitHub repositories based on a given repository name.
 * On success, it dispatches a success action with the fetched data.
 * On failure, it dispatches a failure action with the error data.
 *
 * @date 01/03/2024 - 14:47:28
 *
 * @param {Object} action - The action object containing the repository name.
 * @yields {Effect} The effect of calling the API, and then either the success or failure action.
 */
export function* getGithubRepos(action) {
  const response = yield call(getRepos, action.repoName);
  const { data, ok } = response;
  if (ok) {
    yield put(successGetGithubRepos(data));
  } else {
    yield put(failureGetGithubRepos(data));
  }
}

/**
 * registering events
 * @date 04/03/2024 - 12:57:49
 *
 * @export
 * @returns {{}}
 */
export default function* homeContainerSaga() {
  yield takeLatest(REQUEST_GET_GITHUB_REPOS, getGithubRepos);
}
