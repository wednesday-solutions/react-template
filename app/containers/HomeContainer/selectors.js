import { createSelector } from 'reselect';
import get from 'lodash/get';
import { initialState } from './reducer';

/**
 * Direct selector to the homeContainer state domain
 */

export const selectHomeContainerDomain = (state) => state.homeContainer || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by HomeContainer
 */

export const selectReposData = () =>
  createSelector(selectHomeContainerDomain, (substate) => get(substate, 'reposData'));

export const selectLoading = () => createSelector(selectHomeContainerDomain, (substate) => get(substate, 'loading'));

export const selectReposError = () =>
  createSelector(selectHomeContainerDomain, (substate) => get(substate, 'reposError'));

export const selectRepoName = () => createSelector(selectHomeContainerDomain, (substate) => get(substate, 'repoName'));
