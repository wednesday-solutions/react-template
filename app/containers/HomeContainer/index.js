import React, { useEffect, memo, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import get from 'lodash/get';
import debounce from 'lodash/debounce';
import isEmpty from 'lodash/isEmpty';
import styled from '@emotion/styled';
import { injectIntl } from 'react-intl';
import { injectSaga } from 'redux-injectors';
import { Card, IconButton, Skeleton, InputAdornment, OutlinedInput } from '@mui/material';
import { Search as SearchIcon } from '@mui/icons-material';
import { useHistory } from 'react-router-dom';
import T from '@components/T';
import If from '@components/If';
import For from '@components/For';
import RepoCard from '@components/RepoCard';
import colors from '@app/themes/colors';
import { selectReposData, selectReposError, selectRepoName } from './selectors';
import { homeContainerCreators } from './reducer';
import homeContainerSaga from './saga';

const CustomCard = styled(Card)`
  && {
    margin: 1.25rem 0;
    padding: 1.25rem;
    max-width: ${(props) => props.maxwidth};
    color: ${(props) => props.color};
    ${(props) => props.color && `color: ${props.color}`};
  }
`;
const Container = styled.div`
  && {
    display: flex;
    flex-direction: column;
    max-width: ${(props) => props.maxwidth}px;
    width: 100%;
    margin: 0 auto;
    padding: ${(props) => props.padding}px;
  }
`;
const RightContent = styled.div`
  display: flex;
  align-self: flex-end;
`;

const StyledT = styled(T)`
  && {
    color: ${colors.gotoStories};
  }
`;

const StyledOutlinedInput = styled(OutlinedInput)`
  legend {
    display: none;
  }
  > fieldset {
    top: 0;
  }
`;

export function HomeContainer({
  dispatchGithubRepos,
  dispatchClearGithubRepos,
  intl,
  reposData,
  reposError,
  repoName,
  maxwidth,
  padding
}) {
  const [loading, setLoading] = useState(false);
  const history = useHistory();

  useEffect(() => {
    const loaded = get(reposData, 'items', null) || reposError;
    if (loaded) {
      setLoading(false);
    }
  }, [reposData]);

  useEffect(() => {
    if (repoName && !reposData?.items?.length) {
      dispatchGithubRepos(repoName);
      setLoading(true);
    }
  }, []);

  const searchRepos = (rName) => {
    dispatchGithubRepos(rName);
    setLoading(true);
  };

  const handleOnChange = (rName) => {
    if (!isEmpty(rName)) {
      searchRepos(rName);
    } else {
      dispatchClearGithubRepos();
    }
  };
  const debouncedHandleOnChange = debounce(handleOnChange, 200);

  const renderSkeleton = () => {
    return (
      <>
        <Skeleton data-testid="skeleton" animation="wave" variant="text" height={40} />
        <Skeleton data-testid="skeleton" animation="wave" variant="text" height={40} />
        <Skeleton data-testid="skeleton" animation="wave" variant="text" height={40} />
      </>
    );
  };

  const renderRepoList = () => {
    const items = get(reposData, 'items', []);
    const totalCount = get(reposData, 'totalCount', 0);
    return (
      <If condition={!isEmpty(items) || loading}>
        <CustomCard>
          <If condition={!loading} otherwise={renderSkeleton()}>
            <>
              <If condition={!isEmpty(repoName)}>
                <div>
                  <T id="search_query" values={{ repoName }} />
                </div>
              </If>
              <If condition={totalCount !== 0}>
                <div>
                  <T id="matching_repos" values={{ totalCount }} />
                </div>
              </If>
              <For
                of={items}
                ParentComponent={Container}
                renderItem={(item, index) => <RepoCard key={index} {...item} />}
              />
            </>
          </If>
        </CustomCard>
      </If>
    );
  };
  const renderErrorState = () => {
    let repoError;
    if (reposError) {
      repoError = reposError;
    } else if (isEmpty(repoName)) {
      repoError = 'repo_search_default';
    }
    return (
      !loading &&
      repoError && (
        <CustomCard color={reposError ? 'red' : 'grey'} title={intl.formatMessage({ id: 'repo_list' })}>
          <If condition={reposError} otherwise={<T data-testid="default-message" id={repoError} />}>
            <T data-testid="error-message" text={reposError} />
          </If>
        </CustomCard>
      )
    );
  };

  const handleStoriesClick = () => {
    history.push('/stories');
    window.location.reload();
  };

  return (
    <Container maxwidth={maxwidth} padding={padding}>
      <RightContent>
        <StyledT onClick={handleStoriesClick} data-testid="redirect" id="stories" />
      </RightContent>
      <CustomCard title={intl.formatMessage({ id: 'repo_search' })} maxwidth={maxwidth}>
        <T marginBottom={10} id="get_repo_details" />

        <StyledOutlinedInput
          inputProps={{ 'data-testid': 'search-bar' }}
          onChange={(event) => debouncedHandleOnChange(event.target.value)}
          fullWidth
          defaultValue={repoName}
          endAdornment={
            <InputAdornment position="end">
              <IconButton
                data-testid="search-icon"
                aria-label="search repos"
                type="button"
                onClick={() => searchRepos(repoName)}
              >
                <SearchIcon />
              </IconButton>
            </InputAdornment>
          }
        />
      </CustomCard>
      {renderRepoList()}
      {renderErrorState()}
    </Container>
  );
}

HomeContainer.propTypes = {
  dispatchGithubRepos: PropTypes.func,
  dispatchClearGithubRepos: PropTypes.func,
  intl: PropTypes.object,
  reposData: PropTypes.shape({
    totalCount: PropTypes.number,
    incompleteResults: PropTypes.bool,
    items: PropTypes.array
  }),
  reposError: PropTypes.string,
  repoName: PropTypes.string,
  history: PropTypes.object,
  maxwidth: PropTypes.number,
  padding: PropTypes.number
};

HomeContainer.defaultProps = {
  maxwidth: 500,
  padding: 20,
  reposData: {},
  reposError: null
};

const mapStateToProps = createStructuredSelector({
  reposData: selectReposData(),
  reposError: selectReposError(),
  repoName: selectRepoName()
});

export function mapDispatchToProps(dispatch) {
  const { requestGetGithubRepos, clearGithubRepos } = homeContainerCreators;
  return {
    dispatchGithubRepos: (repoName) => dispatch(requestGetGithubRepos(repoName)),
    dispatchClearGithubRepos: () => dispatch(clearGithubRepos())
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  injectIntl,
  withConnect,
  memo,
  injectSaga({ key: 'homeContainer', saga: homeContainerSaga })
)(HomeContainer);

export const HomeContainerTest = compose(injectIntl)(HomeContainer);
