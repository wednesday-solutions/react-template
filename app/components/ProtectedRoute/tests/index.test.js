import React from 'react';
import { renderProvider } from '@utils/testUtils';
import ProtectedRoute from '../index';
import '@testing-library/jest-dom';
import HomeContainer from '@app/containers/HomeContainer';

jest.mock('@utils/routeConstants', () => {
  return {
    dashboard: {
      route: '/',
      isProtected: true
    },
    login: {
      route: '/login',
      isProtected: false
    }
  };
});

describe('<ProtectedRoute />', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(
      <ProtectedRoute isLoggedIn={true} component={HomeContainer} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should  render the component if user logged in and access protected route', () => {
    const { getByTestId } = renderProvider(
      <ProtectedRoute isLoggedIn={true} component={HomeContainer} />
    );
    expect(getByTestId('redirect')).toBeTruthy();
  });

  it('should not render component if user is not logged in', () => {
    renderProvider(
      <ProtectedRoute isLoggedIn={false} component={HomeContainer} handleLogout={submitSpy} />
    );
    expect(submitSpy).toHaveBeenCalledTimes(1);
  });

  it('should render component , not logged in, unprotected route', () => {
    const { getByTestId } = renderProvider(
      <ProtectedRoute isLoggedIn={false} component={HomeContainer} />
    );
    expect(getByTestId('redirect')).toBeTruthy();
  });

  it('should redirect to the dashboard if logged in and accessing login page(unprotected)', () => {
    // This test should verify that the component renders a Navigate component
    // In React Router v6, we can't easily test history navigation in unit tests
    // The important thing is that the component renders without errors
    const { container } = renderProvider(
      <ProtectedRoute isLoggedIn={true} component={HomeContainer} />
    );
    expect(container).toBeInTheDocument();
  });
});
