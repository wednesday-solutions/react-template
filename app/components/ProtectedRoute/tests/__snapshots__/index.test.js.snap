// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProtectedRoute /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="HomeContainer__Container-l0s8pp-1 iQEnjC"
    >
      <div
        class="HomeContainer__RightContent-l0s8pp-2 iWJscD"
      >
        <p
          class="T__StyledText-gjlic1-0 jOByUY HomeContainer__StyledT-l0s8pp-3 iOEXgS"
          data-testid="redirect"
        >
          Go to Storybook
        </p>
      </div>
      <div
        class="ant-card ant-card-bordered HomeContainer__CustomCard-l0s8pp-0 liNnZz css-dev-only-do-not-override-jy3e2f"
        maxwidth="500"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Repository Search
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        >
          <p
            class="T__StyledText-gjlic1-0 gGweRc"
            data-testid="t"
          >
            Get details of repositories
          </p>
          <span
            class="ant-input-group-wrapper ant-input-search css-dev-only-do-not-override-jy3e2f"
          >
            <span
              class="ant-input-wrapper ant-input-group css-dev-only-do-not-override-jy3e2f"
            >
              <input
                class="ant-input css-dev-only-do-not-override-jy3e2f"
                data-testid="search-bar"
                type="text"
                value=""
              />
              <span
                class="ant-input-group-addon"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-jy3e2f ant-btn-default ant-btn-icon-only ant-input-search-button"
                  type="button"
                >
                  <span
                    aria-label="search"
                    class="anticon anticon-search"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="search"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
          </span>
        </div>
      </div>
      <div
        class="ant-card ant-card-bordered HomeContainer__CustomCard-l0s8pp-0 fCikHd css-dev-only-do-not-override-jy3e2f"
        color="grey"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Repository List
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        >
          <p
            class="T__StyledText-gjlic1-0 jOByUY"
            data-testid="default-message"
          >
            Search for a repository by entering it's name in the search box
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
